sw/device/tests/otbn_mem_scramble_test.c

```diff
@@ -0,0 +1,101 @@
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+OTBN_DECLARE_APP_SYMBOLS(randomness);
+static const otbn_app_t kOtbnAppCfiTest = OTBN_APP_T_INIT(randomness);
+
+const test_config_t kTestConfig;
+
+/**
+ * Check whether the content of an otbn memory match the reference data pointed
+ * by `addr`.
+ *
+ * @param ctx The otbn context object.
+ * @param addr The address to the reference data.
+ * @param mem_size The size to the reference data.
+ * @param read_function Pointer to the function to read the data. It can be
+ * either `dif_otbn_imem_read` or `dif_otbn_dmem_read`.
+ * @return The result of the operation.
+ */
+static otbn_result_t otbn_check_mem(
+    otbn_t *ctx, const uint8_t *addr, size_t mem_size,
+    dif_result_t (*read_function)(const dif_otbn_t *otbn, uint32_t offset_bytes,
+                                  void *dest, size_t len_bytes)) {
+  enum { kBufSize = 256 };
+  uint8_t local_buf[kBufSize] = {0};
+  size_t offset = 0;
+  do {
+    size_t remainder = (mem_size - offset);
+    size_t read_size = remainder < kBufSize ? remainder : kBufSize;
+    if (read_function(&ctx->dif, offset, local_buf, read_size) != kDifOk) {
+      return kOtbnError;
+    }
+
+    if (memcmp(addr + offset, local_buf, read_size)) {
+      return kOtbnError;
+    }
+    offset += read_size;
+  } while (offset < mem_size);
+  return kOtbnOk;
+}
+
+/**
+ * Check that the application is loaded correctly to the IMEM and DMEM.
+ *
+ * @param ctx The otbn context object.
+ * @param app The application to load into OTBN.
+ * @return The result of the operation.
+ */
+static otbn_result_t otbn_check_app(otbn_t *ctx, const otbn_app_t app) {
+  const size_t imem_size = app.imem_end - app.imem_start;
+  const size_t data_size = app.dmem_data_end - app.dmem_data_start;
+
+  // Memory images and offsets must be multiples of 32b words.
+  if (imem_size % sizeof(uint32_t) != 0) {
+    return kOtbnBadArg;
+  }
+
+  // Check the IMEM content.
+  otbn_result_t res =
+      otbn_check_mem(ctx, ctx->app.imem_start, imem_size, dif_otbn_imem_read);
+  if (res != kOtbnOk) {
+    return kOtbnError;
+  }
+
+  // Check the DMEM content.
+  return otbn_check_mem(ctx, ctx->app.dmem_data_start, data_size,
+                        dif_otbn_dmem_read);
+}
+
+bool test_main() {
+  // Initialize.
+  otbn_t otbn_ctx;
+  mmio_region_t addr = mmio_region_from_addr(TOP_EARLGREY_OTBN_BASE_ADDR);
+  CHECK(otbn_init(&otbn_ctx, addr) == kOtbnOk);
+
+  /**
+   * Write and read-check OTBN and IMEM for consistency.
+   */
+  CHECK(otbn_load_app(&otbn_ctx, kOtbnAppCfiTest) == kOtbnOk);
+  CHECK(otbn_check_app(&otbn_ctx, kOtbnAppCfiTest) == kOtbnOk);
+
+  /**
+   * Fetch a new key from the OTP_CTRL and ensure that previous contents in the
+   * IMEM and DMEM cannot be read anymore.
+   */
+  CHECK_DIF_OK(dif_otbn_write_cmd(&otbn_ctx.dif, kDifOtbnCmdSecWipeImem));
+  CHECK(otbn_busy_wait_for_done(&otbn_ctx) == kOtbnOk);
+  CHECK(otbn_check_app(&otbn_ctx, kOtbnAppCfiTest) == kOtbnError);
```

Comment:
This last bit looks a bit odd: I would have thought that each 32-bit read from OTBN's memories will give an ECC error with probability 127/128. So I would have expected to end up in a fault handler here. What am I missing?

Reply:
Sorry, I'm not following. Do you expect the ECC error because the memory has been wiped?

Reply:
"Wiping" the memory is done by rotating scrambling keys. So the actual bits stored in the memory are left unchanged, but we now try to unscramble them with the wrong key which (assuming we've not messed up the cryptography...) essentially gives us a random 39-bit value. The top 7 bits of this value are an ECC checksum and we'll get an integrity error unless they happen to be the right 7 bit number, which should happen one time in 128.

Reply:
In particular, I would have expected the call to `otbn_read` inside `otbn_check_mem` to have failed. Clearly it doesn't! So I've either misunderstood something or you've just found an RTL bug... (My guess is the first)

Reply:
I haven't studied how the ECC works, but I'm assuming it does not take in consideration the key, it is computed after the encryption. So the "Cipher text" is still matching with the ECC checksum. However we should get the wrong "Plain text" for decrypting with the wrong key.
Does that make any sense?

Reply:
Maybe I understood wrong, but the addresses are scrambled as mentioned in the OTBN doc "[The data is bound to the address and scrambled before being stored in memory. The addresses are randomly remapped](https://docs.opentitan.org/hw/ip/otbn/doc/index.html#design-details#memory-scrambling)". In this case would the data + ECC not be affected by a new key?

Reply:
What I said above is correct, and your read will have invalid ECC bits. However, it seems that we weren't actually doing anything about this on the Ibex side, relying entirely on the alert handler (which isn't configured in this test).

But Greg has just merged #11001, which should cause Ibex to lock up a couple of cycles after doing this read. Please could you rebase this commit and re-run the test? I *think* it should then fail. I'm also not sure of the best way to get it working again.

@GregAC: We were talking about fatal exception handlers but I've just realised that they won't do much to help if we disable fetch... Can you think of a way to make this test work without a DV component?